sdk/src/test/java/com/akkaserverless/javasdk/eventsourcedentity/CartHandler.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.javasdk.eventsourcedentity;
+
+import com.akkaserverless.javasdk.impl.eventsourcedentity.AbstractEventSourcedEntityHandler;
+import com.example.shoppingcart.ShoppingCartApi;
+import com.example.shoppingcart.domain.ShoppingCartDomain;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+/** Generated, does the routing from command name to concrete method */
+final class CartHandler
+    extends AbstractEventSourcedEntityHandler<ShoppingCartDomain.Cart, CartEntity> {
+
+  public CartHandler(CartEntity entity) {
+    super(entity);
+  }
+
+  @Override
+  public ShoppingCartDomain.Cart handleEvent(ShoppingCartDomain.Cart state, Object event) {
+    if (event instanceof ShoppingCartDomain.ItemAdded) {
+      return entity().itemAdded(state, (ShoppingCartDomain.ItemAdded) event);
+    } else if (event instanceof ShoppingCartDomain.ItemRemoved) {
+      return entity().itemRemoved(state, (ShoppingCartDomain.ItemRemoved) event);
+    } else {
+      throw new IllegalArgumentException("Unknown event type [" + event.getClass() + "]");
+    }
+  }
+
+  @Override
+  public EventSourcedEntityBase.Effect<?> handleCommand(
+      String commandName, ShoppingCartDomain.Cart state, Any command) {
+    try {
+      switch (commandName) {
+        case "AddItem":
+          // FIXME could parsing to the right type also be pulled out of here?
+          return entity().addItem(state, ShoppingCartApi.AddLineItem.parseFrom(command.getValue()));
```

Comment:
I was looking into that yesterday as well. 

If handleCommand gets an `Object command` instead of `Any`, we could check if it's already parsed or not. 

Each command type would need a generated companion, something like: 

```java
class AddLineItemComp {
  public static AddLineItem cast(Object cmd) {
	if (cmd instance Any) 
		return ShoppingCartApi.AddLineItem.parseFrom(((Any) cmd).getValue()));
	else if (cmd instance AddLineItem) 
		return (AddLineItem) cmd;
	else throw new RuntimeException(...);
  }
}
```

Or maybe something like: 

```java
class ShoppingCartConversions {
  public static AddLineItem toAddLineItem(Object cmd);
}
```

That would also allow use to support other formats, like Jsonable or whatever.

Reply:
I'm not sure we would need the generated companion either, I think that the command contains the right information to deserialize into the right type already, just that the typesystem won't know it. So we should be able to deserialize elsewhere and get the right kind of object (but looks like `AnyRef` to us/the typesystem) then we can expect that to be right here and just cast it to the right type here.

Reply:
So for the record, the point of pulling that elsewhere is that for the testkit we can just send the command object and not first have to do any serialization trixery.

Reply:
To pull it elsewhere, we will need to generate another class that parse `Any` to the right type and then pass to the handleCommand, right? 

And if have two variants of handleCommand, one accepting proto Any and another accepting Object? 

Would that be sufficient?

Reply:
I think we should have one logic that works for both, and that it should be to do serialization/deser in the "impl" and then deal with "normal" objects the rest of the way. My impression of having it this close to the actual logic is that it was done like that to support reflection/annotation magics with no parameter, json, primitives and whatnot but now we are aiming for simplicity rather than that ultimate flexibility.

I'm trying out doing this right now, but it turns out the any support doesn't have all types I thought it would. Investigating if I can get around that, hand it more descriptors or something.

Reply:
I got away with doing that, what was needed was passing both (all) the needed proto descriptors to `AnySupport` when creating it, so in most event sourcing cases both the `API` and the `Domain` proto descriptor. Since we are anyway going to codegen that part (I think at least) that shouldn't be a problem.